src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,520 @@
+    // Get all the input pos lists so that we only have to pointer cast the columns once
+    auto input_pos_lists = std::vector<std::shared_ptr<const PosList>>();
+    for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); ++chunk_id) {
+      auto b_column = input_table->get_chunk(chunk_id)->get_column(column_id);
+      auto r_column = std::dynamic_pointer_cast<const ReferenceColumn>(b_column);
+      input_pos_lists.push_back(r_column->pos_list());
+    }
+
+    // Get the row ids that are referenced
+    auto new_pos_list = std::make_shared<PosList>();
+    for (const auto& row : *pos_list) {
+      if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
+        new_pos_list->push_back(RowID{INVALID_CHUNK_ID, INVALID_CHUNK_OFFSET});
+      } else {
+        new_pos_list->push_back((*input_pos_lists[row.chunk_id])[row.chunk_offset]);
+      }
+    }
+
+    return new_pos_list;
+  }
+
+ public:
+  /**
+  * Executes the MPSMJoin operator.
+  **/
+  std::shared_ptr<const Table> _on_execute() {
+    bool include_null_left = (_mode == JoinMode::Left || _mode == JoinMode::Outer);
+    bool include_null_right = (_mode == JoinMode::Right || _mode == JoinMode::Outer);
+    auto radix_clusterer =
+        RadixClusterSortNUMA<T>(_mpsm_join.input_table_left(), _mpsm_join.input_table_right(), _mpsm_join._column_ids,
+                                include_null_left, include_null_right, _cluster_count);
+    // Sort and cluster the input tables
+    auto sort_output = radix_clusterer.execute();
+    _sorted_left_table = std::move(sort_output.clusters_left);
+    _sorted_right_table = std::move(sort_output.clusters_right);
+    _null_rows_left = std::move(sort_output.null_rows_left);
+    _null_rows_right = std::move(sort_output.null_rows_right);
+
+    // this generates the actual join results and fills the _output_pos_lists
+    _perform_join();
+
+    // merge the pos lists into single pos lists
+    auto output_left = _concatenate_pos_lists(_output_pos_lists_left);
+    auto output_right = _concatenate_pos_lists(_output_pos_lists_right);
+
+    // Add the outer join rows which had a null value in their join column
+    if (include_null_left) {
+      for (auto row_id_left : *_null_rows_left) {
+        output_left->push_back(row_id_left);
+        output_right->push_back(NULL_ROW_ID);
+      }
+    }
+    if (include_null_right) {
+      for (auto row_id_right : *_null_rows_right) {
+        output_left->push_back(NULL_ROW_ID);
+        output_right->push_back(row_id_right);
+      }
+    }
+
+    // Add the columns from both input tables to the output
+    ChunkColumns output_columns;
+    _add_output_columns(output_columns, _mpsm_join.input_table_left(), output_left);
+    _add_output_columns(output_columns, _mpsm_join.input_table_right(), output_right);
+
+    // Build the output_table with one Chunk
+    auto output_column_definitions = concatenated(_mpsm_join.input_table_left()->column_definitions(),
```

Comment:
Since it performs an action, I would prefer `concatenate()`

Reply:
I agree and changed it. This is shared by hash join, sort merge join and us, so the change will be reflected there as well.